```rust
error[E0277]: the trait bound `mozjs_sys::jsgc::Heap<*mut js::jsapi::JSObject>: malloc_size_of::MallocSizeOf` is not satisfied
  --> components/script/dom/xrrigidtransform.rs:22:1
   |
22 | #[dom_struct]
   | ^^^^^^^^^^^^^ the trait `malloc_size_of::MallocSizeOf` is not implemented for `mozjs_sys::jsgc::Heap<*mut js::jsapi::JSObject>`
   |
   = note: required by `malloc_size_of::MallocSizeOf::size_of`
error: aborting due to previous error
For more information about this error, try `rustc --explain E0277`.
error: Could not compile `script`.
To learn more, run the command again with --verbose.
The command "./mach test-unit" exited with 101.
```

--- a/components/malloc_size_of/Cargo.toml
+++ b/components/malloc_size_of/Cargo.toml
@@ -14,7 +14,6 @@ servo = [
     "hyper",
     "hyper_serde",
     "keyboard-types",
-    "mozjs",
     "serde",
     "serde_bytes",
     "string_cache",
@@ -33,7 +32,6 @@ hashglobe = { path = "../hashglobe" }
 hyper = { version = "0.12", optional = true }
 hyper_serde = { version = "0.9", optional = true }
 keyboard-types = {version = "0.4.3", optional = true}
-mozjs = { version = "0.10.0", optional = true}
 selectors = { path = "../selectors" }
 serde = { version = "1.0.27", optional = true }
 serde_bytes = { version = "0.10", optional = true }

--- a/components/malloc_size_of/lib.rs
+++ b/components/malloc_size_of/lib.rs
@@ -58,8 +58,6 @@ extern crate hyper;
 extern crate hyper_serde;
 #[cfg(feature = "servo")]
 extern crate keyboard_types;
-#[cfg(feature = "servo")]
-extern crate mozjs as js;
 extern crate selectors;
 #[cfg(feature = "servo")]
 extern crate serde;
@@ -793,15 +791,6 @@ impl MallocSizeOf for Void {
 
 #[cfg(feature = "servo")]
 impl<Static: string_cache::StaticAtomSet> MallocSizeOf for string_cache::Atom<Static> {
-    fn size_of(&self, _ops: &mut MallocSizeOfOps) -> usize {
-        0
-    }
-}
-
-// This is measured properly by the heap measurement implemented in
-// SpiderMonkey.
-#[cfg(feature = "servo")]
-impl<T: Copy + js::rust::GCMethods> MallocSizeOf for js::jsapi::Heap<T> {
     fn size_of(&self, _ops: &mut MallocSizeOfOps) -> usize {
         0
     }

--- a/components/script/dom/audiobuffer.rs
+++ b/components/script/dom/audiobuffer.rs
@@ -42,6 +42,7 @@ type JSAudioChannel = Heap<*mut JSObject>;
 pub struct AudioBuffer {
     reflector_: Reflector,
     /// Float32Arrays returned by calls to GetChannelData.
+    #[ignore_malloc_size_of = "mozjs"]
     js_channels: DomRefCell<Vec<JSAudioChannel>>,
     /// Aggregates the data from js_channels.
     /// This is Some<T> iff the buffers in js_channels are detached.

--- a/components/script/dom/customelementregistry.rs
+++ b/components/script/dom/customelementregistry.rs
@@ -719,7 +719,7 @@ pub enum CustomElementReaction {
     Upgrade(#[ignore_malloc_size_of = "Rc"] Rc<CustomElementDefinition>),
     Callback(
         #[ignore_malloc_size_of = "Rc"] Rc<Function>,
-        Box<[Heap<JSVal>]>,
+        #[ignore_malloc_size_of = "mozjs"] Box<[Heap<JSVal>]>,
     ),
 }
 

--- a/components/script/dom/extendablemessageevent.rs
+++ b/components/script/dom/extendablemessageevent.rs
@@ -24,6 +24,7 @@ use servo_atoms::Atom;
 #[dom_struct]
 pub struct ExtendableMessageEvent {
     event: ExtendableEvent,
+    #[ignore_malloc_size_of = "mozjs"]
     data: Heap<JSVal>,
     origin: DOMString,
     lastEventId: DOMString,

--- a/components/script/dom/filereader.rs
+++ b/components/script/dom/filereader.rs
@@ -83,7 +83,7 @@ pub enum FileReaderReadyState {
 
 #[derive(JSTraceable, MallocSizeOf)]
 pub enum FileReaderResult {
-    ArrayBuffer(Heap<JSVal>),
+    ArrayBuffer(#[ignore_malloc_size_of = "mozjs"] Heap<JSVal>),
     String(DOMString),
 }
 

--- a/components/script/dom/gamepad.rs
+++ b/components/script/dom/gamepad.rs
@@ -32,6 +32,7 @@ pub struct Gamepad {
     connected: Cell<bool>,
     timestamp: Cell<f64>,
     mapping_type: String,
+    #[ignore_malloc_size_of = "mozjs"]
     axes: Heap<*mut JSObject>,
     buttons: Dom<GamepadButtonList>,
     pose: Option<Dom<VRPose>>,

--- a/components/script/dom/globalscope.rs
+++ b/components/script/dom/globalscope.rs
@@ -145,6 +145,7 @@ pub struct GlobalScope {
     /// they're consumed before it'd be reported.
     ///
     /// <https://html.spec.whatwg.org/multipage/#about-to-be-notified-rejected-promises-list>
+    #[ignore_malloc_size_of = "mozjs"]
     uncaught_rejections: DomRefCell<Vec<Box<Heap<*mut JSObject>>>>,
 
     /// Promises in this list have previously been reported as rejected
@@ -152,6 +153,7 @@ pub struct GlobalScope {
     /// in the last turn of the event loop.
     ///
     /// <https://html.spec.whatwg.org/multipage/#outstanding-rejected-promises-weak-set>
+    #[ignore_malloc_size_of = "mozjs"]
     consumed_rejections: DomRefCell<Vec<Box<Heap<*mut JSObject>>>>,
 }
 

--- a/components/script/dom/history.rs
+++ b/components/script/dom/history.rs
@@ -40,6 +40,7 @@ enum PushOrReplace {
 pub struct History {
     reflector_: Reflector,
     window: Dom<Window>,
+    #[ignore_malloc_size_of = "mozjs"]
     state: Heap<JSVal>,
     state_id: Cell<Option<HistoryStateId>>,
 }

--- a/components/script/dom/imagedata.rs
+++ b/components/script/dom/imagedata.rs
@@ -25,6 +25,7 @@ pub struct ImageData {
     reflector_: Reflector,
     width: u32,
     height: u32,
+    #[ignore_malloc_size_of = "mozjs"]
     data: Heap<*mut JSObject>,
 }
 

--- a/components/script/dom/messageevent.rs
+++ b/components/script/dom/messageevent.rs
@@ -25,6 +25,7 @@ use std::ptr::NonNull;
 #[dom_struct]
 pub struct MessageEvent {
     event: Event,
+    #[ignore_malloc_size_of = "mozjs"]
     data: Heap<JSVal>,
     origin: DOMString,
     source: Option<Dom<WindowProxy>>,

--- a/components/script/dom/paintworkletglobalscope.rs
+++ b/components/script/dom/paintworkletglobalscope.rs
@@ -76,6 +76,7 @@ pub struct PaintWorkletGlobalScope {
     /// <https://drafts.css-houdini.org/css-paint-api/#paint-definitions>
     paint_definitions: DomRefCell<HashMap<Atom, Box<PaintDefinition>>>,
     /// <https://drafts.css-houdini.org/css-paint-api/#paint-class-instances>
+    #[ignore_malloc_size_of = "mozjs"]
     paint_class_instances: DomRefCell<HashMap<Atom, Box<Heap<JSVal>>>>,
     /// The most recent name the worklet was called with
     cached_name: DomRefCell<Atom>,
@@ -473,7 +474,9 @@ pub enum PaintWorkletTask {
 #[derive(JSTraceable, MallocSizeOf)]
 #[must_root]
 struct PaintDefinition {
+    #[ignore_malloc_size_of = "mozjs"]
     class_constructor: Heap<JSVal>,
+    #[ignore_malloc_size_of = "mozjs"]
     paint_function: Heap<JSVal>,
     constructor_valid_flag: Cell<bool>,
     context_alpha_flag: bool,

--- a/components/script/dom/vreyeparameters.rs
+++ b/components/script/dom/vreyeparameters.rs
@@ -22,6 +22,7 @@ pub struct VREyeParameters {
     reflector_: Reflector,
     #[ignore_malloc_size_of = "Defined in rust-webvr"]
     parameters: DomRefCell<WebVREyeParameters>,
+    #[ignore_malloc_size_of = "mozjs"]
     offset: Heap<*mut JSObject>,
     fov: Dom<VRFieldOfView>,
 }

--- a/components/script/dom/vrframedata.rs
+++ b/components/script/dom/vrframedata.rs
@@ -22,9 +22,13 @@ use webvr_traits::WebVRFrameData;
 #[dom_struct]
 pub struct VRFrameData {
     reflector_: Reflector,
+    #[ignore_malloc_size_of = "mozjs"]
     left_proj: Heap<*mut JSObject>,
+    #[ignore_malloc_size_of = "mozjs"]
     left_view: Heap<*mut JSObject>,
+    #[ignore_malloc_size_of = "mozjs"]
     right_proj: Heap<*mut JSObject>,
+    #[ignore_malloc_size_of = "mozjs"]
     right_view: Heap<*mut JSObject>,
     pose: Dom<VRPose>,
     timestamp: Cell<f64>,

--- a/components/script/dom/vrpose.rs
+++ b/components/script/dom/vrpose.rs
@@ -17,11 +17,17 @@ use webvr_traits::webvr;
 #[dom_struct]
 pub struct VRPose {
     reflector_: Reflector,
+    #[ignore_malloc_size_of = "mozjs"]
     position: Heap<*mut JSObject>,
+    #[ignore_malloc_size_of = "mozjs"]
     orientation: Heap<*mut JSObject>,
+    #[ignore_malloc_size_of = "mozjs"]
     linear_vel: Heap<*mut JSObject>,
+    #[ignore_malloc_size_of = "mozjs"]
     angular_vel: Heap<*mut JSObject>,
+    #[ignore_malloc_size_of = "mozjs"]
     linear_acc: Heap<*mut JSObject>,
+    #[ignore_malloc_size_of = "mozjs"]
     angular_acc: Heap<*mut JSObject>,
 }
 

--- a/components/script/dom/vrstageparameters.rs
+++ b/components/script/dom/vrstageparameters.rs
@@ -21,6 +21,7 @@ pub struct VRStageParameters {
     reflector_: Reflector,
     #[ignore_malloc_size_of = "Defined in rust-webvr"]
     parameters: DomRefCell<WebVRStageParameters>,
+    #[ignore_malloc_size_of = "mozjs"]
     transform: Heap<*mut JSObject>,
 }
 

--- a/components/script/dom/xmlhttprequest.rs
+++ b/components/script/dom/xmlhttprequest.rs
@@ -137,6 +137,7 @@ pub struct XMLHttpRequest {
     response_type: Cell<XMLHttpRequestResponseType>,
     response_xml: MutNullableDom<Document>,
     response_blob: MutNullableDom<Blob>,
+    #[ignore_malloc_size_of = "mozjs"]
     response_arraybuffer: Heap<*mut JSObject>,
     #[ignore_malloc_size_of = "Defined in rust-mozjs"]
     response_json: Heap<JSVal>,

--- a/components/script/dom/xrrigidtransform.rs
+++ b/components/script/dom/xrrigidtransform.rs
@@ -27,6 +27,7 @@ pub struct XRRigidTransform {
     #[ignore_malloc_size_of = "defined in euclid"]
     transform: RigidTransform3D<f64>,
     inverse: MutNullableDom<XRRigidTransform>,
+    #[ignore_malloc_size_of = "defined in mozjs"]
     matrix: Heap<*mut JSObject>,
 }
 

--- a/components/script/dom/xrview.rs
+++ b/components/script/dom/xrview.rs
@@ -21,7 +21,9 @@ pub struct XRView {
     reflector_: Reflector,
     session: Dom<XRSession>,
     eye: XREye,
+    #[ignore_malloc_size_of = "mozjs"]
     proj: Heap<*mut JSObject>,
+    #[ignore_malloc_size_of = "mozjs"]
     view: Heap<*mut JSObject>,
     transform: Dom<XRRigidTransform>,
 }

--- a/components/script/dom/xrviewerpose.rs
+++ b/components/script/dom/xrviewerpose.rs
@@ -22,6 +22,7 @@ use webvr_traits::WebVRFrameData;
 #[dom_struct]
 pub struct XRViewerPose {
     pose: XRPose,
+    #[ignore_malloc_size_of = "mozjs"]
     views: Heap<JSVal>,
 }
 

